man/singularity.1
.TH SINGULARITY 1
.SH NAME
Singularity \- Linux Application Containers
.SH SYNOPSIS
.B singularity
[global options...]
.B <command>
[command options...]
.B ...
.SH DESCRIPTION
Singularity provides an application virtualization layer enabling mobility of
compute via both application and environment portability. With Singularity
one is capable of building a root file system and running that root file
system on any other Linux system where Singularity is installed.
.SH OPTIONS
.IP "-h/--help/help"
Print usage summary
.IP "--version"
Print Singularity version
.IP "-v/--verbose"
For each pass of verbose, increase verbosity
.IP "-d/--debug"
Print LOTS of debugging output (including user and process information)
.SH FILES
.I ${sysconfdir}/singularity/singularity.conf
.RS
The system wide configuration file. This file must be root owned for
Singularity to parse it.
.SH EXAMPLES
.IP "$ singularity help"
Will print a generalized usage summary and available commands.
.IP "$ singularity help <command>"
Additional help for any of the Singularity commands can be seen by appending
the command name to the above command.
.SH DOCUMENTATION
Full documentation can be found on the Singularity web site at:
.B https://www.sylabs.io/docs/
.SH BUGS
Please submit bug reports to the GitHub issue tracker at
.B https://github.com/singularityware/singularity/issues/new
.SH AUTHOR
Gregory M. Kurtzer <gmkurtzer@lbl.gov>


CHANGELOG.md
# CHANGELOG

This is a manually generated log to track changes to the repository for each release. 
Each section should include general headers such as ### Implemented enhancements 
and **Merged pull requests**. All closed issued and bug fixes should be 
represented by the pull requests that fixed them. This log originated with Singularity 2.4
and changes prior to that are (unfortunately) done retrospectively. Critical items to know are:

 - renamed, deprecated, or removed commands
 - defaults that are changed
 - backward incompatible changes (recipe file format? image file format?)
 - migration guidance (how to convert images?)
 - changed behaviour (recipe sections work differently)

## [v2.6.0]
    
### Implemented enhancements
 - Allow admin to specify a non-standard location for mksquashfs binary at 
   build time with `--with-mksquashfs` option #1662
 - `--nv` option will use [nvidia-container-cli](https://github.com/NVIDIA/libnvidia-container) if installed #1681
 - [nvliblist.conf](https://github.com/singularityware/singularity/blob/master/etc/nvliblist.conf) now has a section for binaries #1681
 - `--nv` can be made default with all action commands in singularity.conf #1681
 - `--nv` can be controlled by env vars `$SINGULARITY_NV` and 
   `$SINGULARITY_NV_OFF` #1681
 - Refactored travis build and packaging tests #1601
 - Added build and packaging tests for Debian 8/9 and openSUSE 42.3/15.0 #1713
 - Restore shim init process for proper signal handling and child reaping when
   container is initiated in its own PID namespace #1221
 - Add `-i` option to image.create to specify the inode ratio. #1759
 - Bind `/dev/nvidia*` into the container when the `--nv` flag is used in 
    conjuction with the `--contain` flag #1358
 - Add `--no-home` option to not mount user $HOME if it is not the $CWD and
   `mount home = yes` is set. #1761
 - Added support for OAUTH2 Docker registries like Azure Container Registry #1622
 - Include /dev/infiniband if it exists when --contain is used #1810

### Bug fixes
 - Fix 404 when using Arch Linux bootstrap #1731
 - Fix environment variables clearing while starting instances #1766

## [v2.5.2](https://github.com/singularityware/singularity/releases/tag/2.5.2) (2018-07-03)

### [Security related fixes](https://cve.mitre.org/cgi-bin/cvename.cgi?name=2018-12021)
 - Removed the option to use overlay images with `singularity mount`.  This 
   flaw could allow a malicious user accessing the host system to access
   sensitive information when coupled with persistent ext3 overlay.
 - Fixed a race condition that might allow a malicious user to bypass directory 
   image restrictions, like mounting the host root filesystem as a container 
   image

### Bug fixes
 - Fix an error in malloc allocation #1620
 - Honor debug flag when pulling from docker hub #1556
 - Fix a bug with passwd abort #1580
 - Allow user to override singularity.conf "mount home = no" with --home option
   #1496
 - Improve debugging output #1535
 - Fix some bugs in bind mounting #1525
 - Define PR_(S|G)ET_NO_NEW_PRIVS in user space so that these features will 
   work with kernels that implement them (like Cray systems) #1506
 - Create /dev/fd and standard streams symlinks in /dev when using minimal dev
   mount or when specifying -c/-C/--contain option #1420
 - Fixed * expansion during app runscript creation #1486

## [v2.5.1](https://github.com/singularityware/singularity/releases/tag/2.5.1) (2018-05-03)

### Bug fixes
 - Corrected a permissions error when attempting to run Singularity from a 
   directory on NFS with root_squash enabled  
 - Fixed a bug that closed a socket early, preventing correct container 
   execution on hosts using identity services like SSSD
 - Fixed a regression that broke the debootstrap agent

## [v2.5.0](https://github.com/singularityware/singularity/releases/tag/2.5.0) (2018-04-27)

### Security related fixes

Patches are provided to prevent a malicious user with the ability to log in to 
the host system and use the Singularity container runtime from carrying out any 
of the following actions:

 - Create world writable files in root-owned directories on the host system by 
   manipulating symbolic links and bind mounts 
 - Create folders outside of the container by manipulating symbolic links in 
   conjunction with the `--nv` option or by bypassing check_mounted function 
   with relative symlinks
 - Bypass the `enable overlay = no` option in the `singularity.conf` 
   configuration file by setting an environment variable
 - Exploit buffer overflows in `src/util/daemon.c` and/or 
   `src/lib/image/ext3/init.c` (reported by Erik Sjölund (DBB, Stockholm 
   University, Sweden))
 - Forge of the pid_path to join any Singularity namespace (reported by Erik 
   Sjölund (DBB, Stockholm University, Sweden))

### Implemented enhancements

 - Restore docker-extract aufs whiteout handling that implements correct
   extraction of docker container layers. This adds libarchive-devel as a
   build time dep. At runtime libarchive is needed for whiteout handling. If
   libarchive is not available at runtime will fall back to previous
   extraction method.
 - Changed behavior of SINGULARITYENV_PATH to overwrite container PATH and
   added SINGULARITYENV_PREPEND_PATH and SINGULARITYENV_APPEND_PATH for users
   wanting to prepend or append to the container PATH at runtime

### Bug fixes

 - Support pulls from the NVIDIA cloud docker registry (fix by Justin Riley, 
   Harvard)
 - Close socket file descriptors in fd_cleanup
 - Fix conflict between `--nv` and `--contain` options
 - Throw errors at build and runtime if NO_NEW_PRIVS is not present and working
 - Reset umask to 0022 at start to corrrect several errors
 - Verify docker layers after download with sha256 checksum
 - Do not make excessive requests for auth tokens to docker registries
 - Fixed stripping whitespaces and empty new lines for the app commands (fix by 
   Rafal Gumienny, Biozentrum, Basel)

CLI/singularity.go
/*
Copyright (c) 2018, Sylabs, Inc. All rights reserved.
This software is licensed under a 3-clause BSD license.  Please
consult LICENSE file distributed with the sources of this project regarding
your rights to use or distribute this software.
*/

package cmd

import (
	"os"

	"github.com/spf13/cobra"
)

// singularity is the base command when called without any subcommands
var singularityCmd = &cobra.Command{Use: "singularity"}

/*
Execute adds all child commands to the root command and sets flags
appropriately.  This is called by main.main(). It only needs to happen once to
the root command (singularity).
*/
func Execute() {
	if err := singularityCmd.Execute(); err != nil {
		os.Exit(1)
	}
}

var verbose bool

func init() {

	singularityCmd.PersistentFlags().BoolP("debug", "d", false, "")
	singularityCmd.PersistentFlags().BoolP("silent", "s", false, "")
	singularityCmd.PersistentFlags().BoolP("quiet", "q", false, "")
	singularityCmd.PersistentFlags().BoolP("verbose", "v", false, "")

	singularityCmd.SetHelpTemplate(`
USAGE: singularity [global options...] <command> [command options...] ...

GLOBAL OPTIONS:
    -d|--debug    Print debugging information
    -h|--help     Display usage summary
    -s|--silent   Only print errors
    -q|--quiet    Suppress all normal output
    --version  Show application version
    -v|--verbose  Increase verbosity +1
    -x|--sh-debug Print shell wrapper debugging information

GENERAL COMMANDS:
    help       Show additional help for a command or container
    selftest   Run some self tests for singularity install

CONTAINER USAGE COMMANDS:
    exec       Execute a command within container
    run        Launch a runscript within container
    shell      Run a Bourne shell within container
    test       Launch a testscript within container

CONTAINER MANAGEMENT COMMANDS:
    apps       List available apps within a container
    bootstrap  *Deprecated* use build instead
    build      Build a new Singularity container
    check      Perform container lint checks
    inspect    Display container's metadata
    mount      Mount a Singularity container image
    pull       Pull a Singularity/Docker container to $PWD

COMMAND GROUPS:
    image      Container image command group
    instance   Persistent instance command group


CONTAINER USAGE OPTIONS:
    see singularity help <command>

For any additional help or support visit the Singularity
website: http://singularity.lbl.gov/
    `)

	singularityCmd.SetUsageTemplate(`
USAGE: singularity [global options...] <command> [command options...] ...
    `)
}
